### PR TITLE
Added GraphQL Sections into Client-side Data Best Practices

### PR DESCRIPTION
**Pull request by timwright12**
Added updated GraphQL information generated by the front-end engineering team.

Also added the client-side data nav item and had to fix a broken item in the gemlock file

### CODE REVIEW

File: _includes/markdown/JavaScript.md [hunk 1]
```diff
@@ -103,7 +103,23 @@ Certain codebases may already have such libraries in place. Many legacy projects
...
+
+An essential part of a GraphQL API is an API schema. GraphQL requires a human-readable schema which describes the types which are available, and how they relate to one another. While writing a schema is reasonably straightforward, utilizing the standardized nature of WordPress’s database could save time. It may well be possible to reuse a schema from other projects, such as generated by a plugin like, [WPGraphQL](https://github.com/wp-graphql/wp-graphql).
+
+#### WPGraphQL
+If you are choosing to use GraphQL on a Wordpress project, it is recommended to use the WPGraphQL plugin. This plugin will make all the data in the WordPress JSON API available through a GraphQL endpoint - in many cases you won’t need to write the schema yourself. This will give you all the benefits of concatenating your data requests as well as easy access to your data as it is output by WordPress. You can retrieve your data by passing a query directly into a simple fetch request:
```
Comment (jasonbahl) [inline _includes/markdown/JavaScript.md]: Hey! Love seeing WPGraphQL in the best practices 😄 

> will make all the data in the WordPress JSON API available through a GraphQL endpoint

Not sure if the "WordPress JSON API" is referring to the WP REST API? If so, this is inaccurate as WPGraphQL does not wrap the WP REST API, it uses its own functions to resolve data and return a response. It _does_ use JSON to return data, per the GraphQL spec, but I don't think I'd refer to `wp_json_encode` as "the WordPress JSON API" 😉

**Comment by timwright12** [comment]
@jasonbahl why you gotta be creepin'? haha, updated! thanks for the feedback!

File: _includes/markdown/JavaScript.md [hunk 2]
```diff
@@ -103,7 +103,23 @@ Certain codebases may already have such libraries in place. Many legacy projects
...
+
+```javascript
+fetch( '/graphql', {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ [insert your query here] }),
```
Comment (jasonbahl) [inline _includes/markdown/JavaScript.md]: So your query will need to look like: 

```
body: JSON.stringify({ query: "[your query string goes here]"})
```

File: _includes/markdown/JavaScript.md [hunk 3]
```diff
@@ -103,7 +103,23 @@ Certain codebases may already have such libraries in place. Many legacy projects
...
+
+An essential part of a GraphQL API is an API schema. GraphQL requires a human-readable schema which describes the types which are available, and how they relate to one another. While writing a schema is reasonably straightforward, utilizing the standardized nature of WordPress’s database could save time. It may well be possible to reuse a schema from other projects, such as generated by a plugin like, [WPGraphQL](https://github.com/wp-graphql/wp-graphql).
+
+#### WPGraphQL
+If you are choosing to use GraphQL on a Wordpress project, it is recommended to use the WPGraphQL plugin. This plugin will return WordPress data in JSON format through a GraphQL endpoint - in many cases you won’t need to write the schema yourself. This will give you all the benefits of concatenating your data requests as well as easy access to your data as it is output by WordPress. You can retrieve your data by passing a query directly into a simple fetch request:
```
Comment (nicholasio) [inline _includes/markdown/JavaScript.md]: change Wordpress to WordPress in `GraphQL on a Wordpress project, it is `

File: _includes/markdown/JavaScript.md [hunk 4]
```diff
@@ -103,7 +103,23 @@ Certain codebases may already have such libraries in place. Many legacy projects
...
+
+```javascript
+fetch( '/graphql', {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ [insert your query here] }),
+} ).then( res => res.json() );
+```
+
+The above code snippet will help you get started in making requests to the GraphQL service.
```
Comment (nicholasio) [inline _includes/markdown/JavaScript.md:122]: I think we should stress out that there are a lot benefits in using a dedicated GraphQL client. if  GraphQL was introduced in a project it likely have a certain level of complexity that is going to be more than a simple one time graphql request.

**Review by nicholasio — DISMISSED** [review]

File: _includes/markdown/JavaScript.md (hunk 2 above)
Comment (jasonbahl) [reply _includes/markdown/JavaScript.md]: with WPGraphQL (or any other GraphQL server that supports Batch requests, you can also send an array of queries like so: 

```
body: JSON.stringify([{ query: "[your query string goes here]"}, {query: "[your next query goes here]"}])
```

Apollo Client can automatically generate and send batch queries for you via Apollo Link Batch HTTP: https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-batch-http. 

Batch queries can further increase the performance of client-side applications as fewer HTTP requests are needed to get the data necessary for your application. 

This is common when first loading a page or changing routes that are composed of many components that each have their own GraphQL dependencies. 

For example, let's say you had the following: 

```
<Layout>
  <Menu />
  <RecentPosts />
</Layout>
```

The `<Menu>` component could make use of a WPGraphQL `menuItems` query, and the `<RecentPosts />` component could make use of a `posts` query. 

When the page first loads, Apollo Link HTTP Batch would send both the `menuItems` query _and_ the `posts` query in a single HTTP request. 

So you get the benefit of declaratively writing queries for the data you need, coupling the data dependencies with the components that need the data, and having the data fetched in the most efficient way possible. 💥

Comment (jasonbahl) [reply _includes/markdown/JavaScript.md]: fwiw, ^ that's just some helpful information. . .up to you if ya'll want to mention it in the README or not. 😄

File: _includes/markdown/JavaScript.md (hunk 4 above)
Comment (timwright12) [reply _includes/markdown/JavaScript.md:122]: Happy to add a blurb if you have something. I definitely don't want it to be a GraphQL tutorial though.

**Pull request merged**